mautic/mautic
# Feature/4922 images in form

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/4922
| BC breaks? | N
| Deprecations? | N

[//]: # ( Required: )
#### Description:
This PR adds the ability to upload a file as a form field so that I can collect and store documents from a contact easily.

#### Steps to test this PR:
1. Create a new form and add a file field.
2. Go to the form, choose a file and send it. You should see a link to file download in results.
3. Go to Contacts, click on your user (maybe send form in anonymous window). There should be link to file in contact details under form submission.
4. Set different validation rules for file field (try required, allowed extensions and file size). Each validation should work. Make sure that if a validation fails, no file is uploaded to `Your-Mautic-Folder/media/files/form/`
5. Maximum file size cannot be greater than limit of your PHP configuration. Limit is shown in the tooltip of this field. Try to set bigger limit and you should get an error.
6. There is a `blacklisted_extensions` option in configuration. Check the tooltip of the field - supported format should be there. Admin should not be able to set forbidden extension to `Allowed file extensions` field. There are 2 extensions in default: `php` and `sh`. Try to allow one of them (and both later) and you should get an error message.
7. Upload a file through the form, check that file was uploaded in your file system and delete the submission from admin. Uploaded file should be deleted.
8. Upload a new file, check it on the file system and delete file field from the form. File should be deleted from file system. Check, that only field's directory was be deleted and other field's folders / files remain.
9. Create new form with several upload fields and send several submissions. Check your file system for files. Delete whole form. All files and directories of the form should be deleted. Check that other directory (of other form) was not modified / deleted.
